client/my-sites/add-ons/hooks/use-add-ons.ts

```diff
@@ -73,10 +75,28 @@ const useAddOns = (): ( AddOnMeta | null )[] => {
 		},
 	];
 
-	// Filter out storage add ons if they are disabled in the config
-	const filteredAddOns = ! isStorageAddonEnabled()
-		? addOnsActive.filter( ( addOn ) => addOn.productSlug !== PRODUCT_1GB_SPACE )
-		: addOnsActive;
+	// try to build storage add ons
+	const { data: mediaStorage } = useMediaStorageQuery( siteId );
+
+	const filteredAddOns = addOnsActive.filter( ( addOn ) => {
+		// if it's a storage add on
+		if ( addOn.productSlug === PRODUCT_1GB_SPACE ) {
+			// if storage add ons are not enabled in the config, remove them
+			if ( ! isStorageAddonEnabled() ) {
+				return false;
+			}
+
+			const currentMaxStorage = mediaStorage?.max_storage_bytes / Math.pow( 1024, 3 );
+			const availableStorageUpgrade = STORAGE_LIMIT - currentMaxStorage;
+
+			// if the current storage add on option is greater than the available upgrade, remove it
+			if ( ( addOn.quantity as number ) > availableStorageUpgrade ) {
```

Comment:
Really minor and not a blocker but this can be done without needing to resort to `as` (in this case it's quite safe but I always like to encourage folks to avoid `as` when possible... and it's not always possible so this is just a suggestion).

```suggestion
			if ( ( addOn.quantity ?? 0 ) > availableStorageUpgrade ) {
```